resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,575 @@
+		return (T t) -> {
+			long start = 0;
+			boolean isFailed=false;
+			bulkhead.acquirePermission();
+			try {
+				start = Instant.now().toEpochMilli();
+				return function.apply(t);
+			} catch (Exception e) {
+				isFailed=true;
+				bulkhead.onError( start,TimeUnit.MILLISECONDS, e);
+				throw e;
+			} finally {
+				if(start != 0 && !isFailed) {
+					bulkhead.onSuccess(start, TimeUnit.MILLISECONDS);
+				}
+			}
+		};
+	}
+
+
+
+	/**
+	 * Create a Bulkhead with a default configuration.
+	 *
+	 * @param name the name of the bulkhead
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead ofDefaults(String name) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, AdaptiveBulkheadConfig.ofDefaults());
+	}
+
+
+	/**
+	 * Creates a bulkhead with a custom configuration and custom limiter
+	 *
+	 * @param name         the name of the bulkhead
+	 * @param config       a custom BulkheadConfig configuration
+	 * @param limitAdapter the custom limit adopter
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config, LimitPolicy limitAdapter) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config, limitAdapter);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name   the name of the bulkhead
+	 * @param config a custom BulkheadConfig configuration
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name                   the name of the bulkhead
+	 * @param bulkheadConfigSupplier custom configuration supplier
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, Supplier<AdaptiveBulkheadConfig> bulkheadConfigSupplier) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, bulkheadConfigSupplier.get());
+	}
+
+	interface AdaptiveMetrics extends Bulkhead.Metrics {
+		/**
+		 * Returns the current total number of calls which were slower than a certain threshold.
+		 *
+		 * @return the current total number of calls which were slower than a certain threshold
+		 */
+		int getNumberOfSlowCalls();
+
+		/**
+		 * Returns the current number of failed buffered calls in the ring buffer.
+		 *
+		 * @return the current number of failed buffered calls in the ring buffer
+		 */
+		int getNumberOfFailedCalls();
+
+		/**
+		 * Returns the current number of successful buffered calls in the ring buffer.
+		 *
+		 * @return the current number of successful buffered calls in the ring buffer
+		 */
+		int getNumberOfSuccessfulCalls();
+
+		/**
+		 * @return average latency for service calls in millis
+		 */
+		double getAverageLatencyMillis();
```

Comment:
Do we need to make `getAverageLatencyMillis` public?

Reply:
in case if community would like to introduced new policies like moving average ..ect , it can be useful for them especially we export that anyhow through our metrics